Spectral OAS2 validation throws a non-existent error without path value

Running spectral OAS2 rules and functions sets on a OAS2 `.yaml` file on all tests (2 different `yaml files`) results in a cryptic error without a `"path"` parameter:
```
{
    "name": "oas2-schema",
    "summary": "Validate structure of OpenAPIv2 specification.",
    "message": "should NOT have additional properties: 0",
    "path": [],
    "severity": 50,
    "severityLabel": "error"
}
```
Those files don't throw any errors in any other validation tool (swagger editor, apiary, etc.). I am happy to provide them if needed.

### **I'm submitting a...**
  - [X] bug report
  - [ ] feature request

### What is the current behavior?

Running:
```
spectral.addFunctions(oas2Functions());
spectral.addRules(oas2Rules());
const results = spectral.run(myOAS);
```
on a OAS2 `.yaml` file

### What is the expected behavior?
Do not throw this error or be more specific on it's origin and place in the file.

### What is the motivation / use case for changing the behavior?
User experience and compliance with OAS2 specification

### Please tell us about your environment:

  - Version: @stoplight/spectral@1.2.0
  - Framework: [ ]
  - Language: [all]


### Other information

It also throws 
```
oasOpParams expects a resolved object, but none was provided. Results may not be correct.
oasPathParam expects a resolved object, but none was provided. Results may not be correct.
```
which does not say anything of value without more explanation.

Hey! So, good news and bad news @jerzyn. I have got the release/2.0 branch CLI taking a look at your spec and it has a _lot_ of opinions.

```bash
$ git checkout release/2.0
$ yarn setup && yarn build
$ ./dist/bin/run lint https://gist.githubusercontent.com/jerzyn/d0201558bf52b3659c0d43cac646896c/raw/0bc6e80d7bce9d8556f791a641924beb624e2f7b/swagger-demo.yaml

OpenAPI 2.0 (Swagger) detected
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id}/reviews,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id}/reviews-stats,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id}/wishlist,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id}/wishlist/{product_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'valid-example' on node at path '$,definitions,halRoot':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halProducts':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halProduct':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,productPatch':
Error: can't resolve reference #/definitions/product from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,productNotFound':
Error: can't resolve reference #/definitions/problemDetail from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviews':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReview':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviewsStats':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviewStatsItem':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halUser':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,userNotFound':
Error: can't resolve reference #/definitions/problemDetail from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halWishlist':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halWishProduct':
Error: can't resolve reference #/definitions/halResource from id #

/Users/phil/src/spectral/https:/gist.githubusercontent.com/jerzyn/d0201558bf52b3659c0d43cac646896c/raw/0bc6e80d7bce9d8556f791a641924beb624e2f7b/swagger-demo.yaml
   1:1   warning  api-host               OpenAPI `host` must be present and non-empty string
   2:6   warning  info-contact           Info object should contain `contact` object
  55:16  warning  operation-tags         Operation should have non-empty `tags` array
  55:16  warning  operation-operationId  Operation should have an `operationId`
  55:16  warning  operation-description  Operation `description` must be present and non-empty string
  57:9   warning  operation-operationId  Operation should have an `operationId`
  57:9   warning  operation-tags         Operation should have non-empty `tags` array
  69:16  warning  operation-tags         Operation should have non-empty `tags` array
  69:16  warning  operation-operationId  Operation should have an `operationId`
  69:16  warning  operation-description  Operation `description` must be present and non-empty string
  71:9   warning  operation-operationId  Operation should have an `operationId`
  71:9   warning  operation-tags         Operation should have non-empty `tags` array
 124:10  warning  operation-operationId  Operation should have an `operationId`
 124:10  warning  operation-tags         Operation should have non-empty `tags` array
 146:16  warning  operation-operationId  Operation should have an `operationId`
 146:16  warning  operation-tags         Operation should have non-empty `tags` array
 146:16  warning  operation-description  Operation `description` must be present and non-empty string
  156:9  warning  operation-operationId  Operation should have an `operationId`
  156:9  warning  operation-tags         Operation should have non-empty `tags` array
 176:11  warning  operation-tags         Operation should have non-empty `tags` array
 176:11  warning  operation-operationId  Operation should have an `operationId`
 199:16  warning  operation-operationId  Operation should have an `operationId`
 199:16  warning  operation-description  Operation `description` must be present and non-empty string
 199:16  warning  operation-tags         Operation should have non-empty `tags` array
  209:9  warning  operation-tags         Operation should have non-empty `tags` array
  209:9  warning  operation-operationId  Operation should have an `operationId`
 236:10  warning  operation-operationId  Operation should have an `operationId`
 236:10  warning  operation-tags         Operation should have non-empty `tags` array
 259:16  warning  operation-tags         Operation should have non-empty `tags` array
 259:16  warning  operation-description  Operation `description` must be present and non-empty string
 259:16  warning  operation-operationId  Operation should have an `operationId`
  269:9  warning  operation-operationId  Operation should have an `operationId`
  269:9  warning  operation-tags         Operation should have non-empty `tags` array
 305:16  warning  operation-description  Operation `description` must be present and non-empty string
 305:16  warning  operation-tags         Operation should have non-empty `tags` array
 305:16  warning  operation-operationId  Operation should have an `operationId`
 307:10  warning  operation-operationId  Operation should have an `operationId`
 307:10  warning  operation-tags         Operation should have non-empty `tags` array
 329:16  warning  operation-operationId  Operation should have an `operationId`
 329:16  warning  operation-description  Operation `description` must be present and non-empty string
 329:16  warning  operation-tags         Operation should have non-empty `tags` array
  339:9  warning  operation-operationId  Operation should have an `operationId`
  339:9  warning  operation-tags         Operation should have non-empty `tags` array
 352:16  warning  operation-tags         Operation should have non-empty `tags` array
 352:16  warning  operation-operationId  Operation should have an `operationId`
 352:16  warning  operation-description  Operation `description` must be present and non-empty string
  362:9  warning  operation-tags         Operation should have non-empty `tags` array
  362:9  warning  operation-operationId  Operation should have an `operationId`
 374:12  warning  operation-tags         Operation should have non-empty `tags` array
 374:12  warning  operation-operationId  Operation should have an `operationId`
 385:16  warning  operation-operationId  Operation should have an `operationId`
 385:16  warning  operation-tags         Operation should have non-empty `tags` array
 385:16  warning  operation-description  Operation `description` must be present and non-empty string
 402:10  warning  operation-operationId  Operation should have an `operationId`
 402:10  warning  operation-tags         Operation should have non-empty `tags` array
 423:12  warning  operation-operationId  Operation should have an `operationId`
 423:12  warning  operation-tags         Operation should have non-empty `tags` array
  744:8  warning  valid-example          Examples must be valid against their defined schema

✖ 58 problems (0 errors, 58 warnings, 0 infos)
```

So the TypeError and Error messages at the top looks like bugs, but it also seems that Spectral will keep on trucking past those bugs in order to provide linting regardless.

If you would like to work with the CLI version for now following the instructions above, you can use those to lint, whilst we get these bugs sorted out over in #111.

Hey! So, good news and bad news @jerzyn. I have got the release/2.0 branch CLI taking a look at your spec and it has a _lot_ of opinions.

```bash
$ git checkout release/2.0
$ yarn setup && yarn build
$ ./dist/bin/run lint https://gist.githubusercontent.com/jerzyn/d0201558bf52b3659c0d43cac646896c/raw/0bc6e80d7bce9d8556f791a641924beb624e2f7b/swagger-demo.yaml

OpenAPI 2.0 (Swagger) detected
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id}/reviews,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id}/reviews-stats,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id}/wishlist,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id}/wishlist/{product_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'valid-example' on node at path '$,definitions,halRoot':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halProducts':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halProduct':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,productPatch':
Error: can't resolve reference #/definitions/product from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,productNotFound':
Error: can't resolve reference #/definitions/problemDetail from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviews':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReview':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviewsStats':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviewStatsItem':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halUser':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,userNotFound':
Error: can't resolve reference #/definitions/problemDetail from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halWishlist':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halWishProduct':
Error: can't resolve reference #/definitions/halResource from id #

/Users/phil/src/spectral/https:/gist.githubusercontent.com/jerzyn/d0201558bf52b3659c0d43cac646896c/raw/0bc6e80d7bce9d8556f791a641924beb624e2f7b/swagger-demo.yaml
   1:1   warning  api-host               OpenAPI `host` must be present and non-empty string
   2:6   warning  info-contact           Info object should contain `contact` object
  55:16  warning  operation-tags         Operation should have non-empty `tags` array
  55:16  warning  operation-operationId  Operation should have an `operationId`
  55:16  warning  operation-description  Operation `description` must be present and non-empty string
  57:9   warning  operation-operationId  Operation should have an `operationId`
  57:9   warning  operation-tags         Operation should have non-empty `tags` array
  69:16  warning  operation-tags         Operation should have non-empty `tags` array
  69:16  warning  operation-operationId  Operation should have an `operationId`
  69:16  warning  operation-description  Operation `description` must be present and non-empty string
  71:9   warning  operation-operationId  Operation should have an `operationId`
  71:9   warning  operation-tags         Operation should have non-empty `tags` array
 124:10  warning  operation-operationId  Operation should have an `operationId`
 124:10  warning  operation-tags         Operation should have non-empty `tags` array
 146:16  warning  operation-operationId  Operation should have an `operationId`
 146:16  warning  operation-tags         Operation should have non-empty `tags` array
 146:16  warning  operation-description  Operation `description` must be present and non-empty string
  156:9  warning  operation-operationId  Operation should have an `operationId`
  156:9  warning  operation-tags         Operation should have non-empty `tags` array
 176:11  warning  operation-tags         Operation should have non-empty `tags` array
 176:11  warning  operation-operationId  Operation should have an `operationId`
 199:16  warning  operation-operationId  Operation should have an `operationId`
 199:16  warning  operation-description  Operation `description` must be present and non-empty string
 199:16  warning  operation-tags         Operation should have non-empty `tags` array
  209:9  warning  operation-tags         Operation should have non-empty `tags` array
  209:9  warning  operation-operationId  Operation should have an `operationId`
 236:10  warning  operation-operationId  Operation should have an `operationId`
 236:10  warning  operation-tags         Operation should have non-empty `tags` array
 259:16  warning  operation-tags         Operation should have non-empty `tags` array
 259:16  warning  operation-description  Operation `description` must be present and non-empty string
 259:16  warning  operation-operationId  Operation should have an `operationId`
  269:9  warning  operation-operationId  Operation should have an `operationId`
  269:9  warning  operation-tags         Operation should have non-empty `tags` array
 305:16  warning  operation-description  Operation `description` must be present and non-empty string
 305:16  warning  operation-tags         Operation should have non-empty `tags` array
 305:16  warning  operation-operationId  Operation should have an `operationId`
 307:10  warning  operation-operationId  Operation should have an `operationId`
 307:10  warning  operation-tags         Operation should have non-empty `tags` array
 329:16  warning  operation-operationId  Operation should have an `operationId`
 329:16  warning  operation-description  Operation `description` must be present and non-empty string
 329:16  warning  operation-tags         Operation should have non-empty `tags` array
  339:9  warning  operation-operationId  Operation should have an `operationId`
  339:9  warning  operation-tags         Operation should have non-empty `tags` array
 352:16  warning  operation-tags         Operation should have non-empty `tags` array
 352:16  warning  operation-operationId  Operation should have an `operationId`
 352:16  warning  operation-description  Operation `description` must be present and non-empty string
  362:9  warning  operation-tags         Operation should have non-empty `tags` array
  362:9  warning  operation-operationId  Operation should have an `operationId`
 374:12  warning  operation-tags         Operation should have non-empty `tags` array
 374:12  warning  operation-operationId  Operation should have an `operationId`
 385:16  warning  operation-operationId  Operation should have an `operationId`
 385:16  warning  operation-tags         Operation should have non-empty `tags` array
 385:16  warning  operation-description  Operation `description` must be present and non-empty string
 402:10  warning  operation-operationId  Operation should have an `operationId`
 402:10  warning  operation-tags         Operation should have non-empty `tags` array
 423:12  warning  operation-operationId  Operation should have an `operationId`
 423:12  warning  operation-tags         Operation should have non-empty `tags` array
  744:8  warning  valid-example          Examples must be valid against their defined schema

✖ 58 problems (0 errors, 58 warnings, 0 infos)
```

So the TypeError and Error messages at the top looks like bugs, but it also seems that Spectral will keep on trucking past those bugs in order to provide linting regardless.

If you would like to work with the CLI version for now following the instructions above, you can use those to lint, whilst we get these bugs sorted out over in #111.

--- a/src/functions/__tests__/schema-path.test.ts
+++ b/src/functions/__tests__/schema-path.test.ts
@@ -1,11 +1,16 @@
 import { schemaPath } from '../schema-path';
 
 function runSchemaPath(target: any, field: string, schemaPathStr: string) {
-  return schemaPath(target, { field, schemaPath: schemaPathStr }, { given: ['$'] }, { given: null, original: null });
+  return schemaPath(
+    target,
+    { field, schemaPath: schemaPathStr },
+    { given: [], target: [] },
+    { given: null, original: null, resolved: target }
+  );
 }
 
 describe('schema', () => {
-  // Check the example field matches the conents of schema
+  // Check the example field matches the contents of schema
   const fieldToCheck = 'example';
   const path = '$.schema';
 
@@ -58,9 +63,7 @@ describe('schema', () => {
 Array [
   Object {
     "message": "should have required property 'url'",
-    "path": Array [
-      "$",
-    ],
+    "path": Array [],
   },
 ]
 `);
@@ -79,9 +82,7 @@ Array [
 Array [
   Object {
     "message": "should match format \\"url\\"",
-    "path": Array [
-      "$",
-    ],
+    "path": Array [],
   },
 ]
 `);

--- a/src/functions/schema-path.ts
+++ b/src/functions/schema-path.ts
@@ -13,24 +13,29 @@ import { schema } from './schema';
 import jp = require('jsonpath');
 
 export const schemaPath: IFunction<ISchemaPathOptions> = (targetVal, opts, paths, otherValues) => {
-  if (!targetVal) return [];
-  let value = targetVal;
+  if (!targetVal || typeof targetVal !== 'object') return [];
 
-  // We want to run this on a specific field, for example: example
-  if (opts.field && typeof targetVal === 'object') {
-    value = targetVal[opts.field];
+  let object = targetVal;
+
+  const { resolved } = otherValues;
+  if (!resolved) {
+    console.warn('schema-path expects a resolved object, but none was provided. Results may not be correct.');
+  } else {
+    // Take the relevant part of the resolved schema
+    object = jp.value(resolved, jp.stringify(['$', ...paths.given]));
   }
 
-  if (!value) return [];
+  // The subsection of the targetVal which contains the good bit
+  const relevantObject = opts.field ? object[opts.field] : object;
+  if (!relevantObject) return [];
 
-  // Hunt in the original targetVal for the schema we are going to apply
-  let schemaObj;
+  // The subsection of the targetValue which contains the schema for us to validate the good bit against
+  let schemaObject;
   try {
-    schemaObj = jp.value(targetVal, opts.schemaPath);
+    schemaObject = jp.value(object, opts.schemaPath);
   } catch (error) {
     console.error(error);
   }
-  if (!schemaObj) return [];
 
-  return schema(value, { schema: schemaObj }, paths, otherValues);
+  return schema(relevantObject, { schema: schemaObject }, paths, otherValues);
 };

--- a/src/rulesets/oas2/__tests__/valid-example.ts
+++ b/src/rulesets/oas2/__tests__/valid-example.ts
@@ -118,6 +118,34 @@ describe('valid-example', () => {
           properties: {
             _links: {
               type: 'object',
+              additionalProperties: {
+                allOf: [
+                  {
+                    $ref: '#/definitions/halLinkObject',
+                  },
+                  {
+                    type: 'array',
+                    items: [
+                      {
+                        $ref: '#/definitions/halLinkObject',
+                      },
+                    ],
+                  },
+                ],
+              },
+            },
+            _embedded: {
+              type: 'object',
+              additionalProperties: true,
+            },
+          },
+        },
+        halLinkObject: {
+          type: 'object',
+          required: ['href'],
+          properties: {
+            href: {
+              type: 'string',
             },
           },
         },
@@ -128,7 +156,7 @@ describe('valid-example', () => {
       expect.objectContaining({
         code: 'valid-example',
         summary: 'Examples must be valid against their defined schema.',
-        path: ['paths', '/path', 'get', 'responses'],
+        path: ['definitions', 'halRoot', '_links', 'self'],
       }),
     ]);
   });

--- a/src/rulesets/oas2/__tests__/valid-example.ts
+++ b/src/rulesets/oas2/__tests__/valid-example.ts
@@ -85,6 +85,54 @@ describe('valid-example', () => {
     expect(results).toHaveLength(1);
   });
 
+  test('works fine with allOf $ref', async () => {
+    const results = await s.run({
+      definitions: {
+        halRoot: {
+          type: 'object',
+          allOf: [
+            {
+              $ref: '#/definitions/halResource',
+            },
+          ],
+          example: {
+            _links: {
+              self: {
+                href: '/',
+              },
+              products: {
+                href: '/products',
+              },
+              product: {
+                href: '/products/{product_id}',
+              },
+              users: {
+                href: '/users',
+              },
+            },
+          },
+        },
+        halResource: {
+          title: 'HAL Resource Object',
+          type: 'object',
+          properties: {
+            _links: {
+              type: 'object',
+            },
+          },
+        },
+      },
+    });
+
+    expect(results).toEqual([
+      expect.objectContaining({
+        code: 'valid-example',
+        summary: 'Examples must be valid against their defined schema.',
+        path: ['paths', '/path', 'get', 'responses'],
+      }),
+    ]);
+  });
+
   test('will pass for valid parents examples which contain invalid child examples', async () => {
     const results = await s.run({
       swagger: '2.0',

